server/constraints.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/juju/errors"
+)
+
+// Constraint is a replica constraint to place region peers.
+type Constraint struct {
+	Labels   map[string]string
+	Replicas int
+}
+
+func (c *Constraint) String() string {
+	return fmt.Sprintf("Constraint(%+v)", *c)
+}
+
+// Match returns true if store matches the constraint.
+func (c *Constraint) Match(store *storeInfo) bool {
+	labels := make(map[string]string)
+	for _, label := range store.GetLabels() {
```

Comment:
why not checking `len(store.GetLabels() == len(c.Labels)` first, then:

```
for _, label := range store.GetLabels() {
    if v := c.Labels[label.GetKey()); v != label.GetValue() {
        return false 
    }
}
```

So we don't need using another map here.